File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""

        need = {}
        window = {}

        for c in t:
            need[c] = need.get(c, 0) + 1

        left = 0
        formed = 0
        required = len(need)

        min_len = float("inf")
        start = 0

        for right in range(len(s)):

            c = s[right]
            window[c] = window.get(c, 0) + 1

            if c in need and window[c] == need[c]:
                formed += 1

            while formed == required:

                if right - left + 1 < min_len:
                    min_len = right - left + 1
                    start = left

                left_char = s[left]
                window[left_char] -= 1

                if (left_char in need and
                        window[left_char] < need[left_char]):
                    formed -= 1

                left += 1

        if min_len == float("inf"):
            return ""

        return s[start:start + min_len]

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t.size() > s.size()) return "";

        unordered_map<char, int> need;
        unordered_map<char, int> window;

        // Frequency of characters required from t
        for (char c : t) {
            need[c]++;
        }

        int left = 0;
        int formed = 0;
        int required = need.size();

        int minLen = INT_MAX;
        int start = 0;

        for (int right = 0; right < s.size(); right++) {

            char c = s[right];
            window[c]++;

            // This character has now satisfied its requirement
            if (need.count(c) && window[c] == need[c]) {
                formed++;
            }

            // Current window contains all characters of t
            while (formed == required) {

                // Update smallest window
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    start = left;
                }

                // Remove left character
                char leftChar = s[left];
                window[leftChar]--;

                // Window is no longer satisfying this character
                if (need.count(leftChar) &&
                    window[leftChar] < need[leftChar]) {
                    formed--;
                }

                left++;
            }
        }

        if (minLen == INT_MAX)
            return "";

        return s.substr(start, minLen);
    }
};